courseworks/week3/xuanzhaogao/hw3.ipynb

```diff
@@ -0,0 +1,192 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "1. Get the relative condition number of division operation $a/b$.\n",
+    "\n",
+    "$$\n",
+    "\\frac{||\\delta\\frac{a}{b}||}{||\\frac{a}{b}||} = \\frac{||\\frac{\\delta a}{b} - \\frac{a \\delta b}{b^2}||}{||\\frac{a}{b}||} = ||\\frac{\\delta a}{a} - \\frac{\\delta b}{b}|| \\leq {\\delta \\epsilon}(||\\frac{1}{a}|| + ||\\frac{1}{b}||)\n",
+    "$$\n",
+    "\n",
+    "\n",
+    "$$\n",
+    "\\kappa(\\frac{a}{b}) = {\\delta \\epsilon}(||\\frac{1}{a}|| + ||\\frac{1}{b}||) / ( {\\delta \\epsilon} / ||a + b||) = \\frac{1}{||ab||}\n",
+    "$$"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "2. Classify each of the following matrices as well-conditioned or ill-conditioned:\n",
+    "```math\n",
+    "(a). ~~\\left(\\begin{matrix}10^{10} & 0\\\\ 0 & 10^{-10}\\end{matrix}\\right)\n",
+    "```\n",
+    "$\\kappa = 10^{20}$, ill-conditioned.\n",
+    "\n",
+    "```math\n",
+    "(b). ~~\\left(\\begin{matrix}10^{10} & 0\\\\ 0 & 10^{10}\\end{matrix}\\right)\n",
+    "```\n",
+    "$\\kappa = 1$, well-conditioned.\n",
+    "\n",
+    "```math\n",
+    "(c). ~~\\left(\\begin{matrix}10^{-10} & 0\\\\ 0 & 10^{-10}\\end{matrix}\\right)\n",
+    "```\n",
+    "$\\kappa = 1$, well-conditioned.\n",
+    "\n",
+    "```math\n",
+    "(d). ~~\\left(\\begin{matrix}1 & 2\\\\ 2 & 4\\end{matrix}\\right)\n",
+    "```\n",
+    "\n",
+    "$\\lambda = 0, 5$, ill-conditioned.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "2-element Vector{Float64}:\n",
+       " 0.0\n",
+       " 5.0"
+      ]
+     },
+     "execution_count": 2,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "using LinearAlgebra\n",
+    "A = [1 2; 2 4]\n",
+    "eigvals(A)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "gauss_jordan (generic function with 1 method)"
+      ]
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "using LinearAlgebra\n",
+    "\n",
+    "function gauss_jordan(A)\n",
+    "    n = size(A,1)\n",
+    "    B = Matrix{Float64}(I, n, n)\n",
+    "    for i = 1:n #i is for the column\n",
+    "        C = Matrix{Float64}(I, n, n)\n",
+    "        for j in 1:n #j is for the row\n",
+    "            C[j, i] = - A[j, i] / A[i, i]\n",
```

Comment:
The implementation of Gauss-Jordan elimination can be improved. e.g. Implementing it as a "linear operator" rather than a matrix. Please check the hoseholder reflextor in lecture-5. https://giggleliu.github.io/ModernScientificComputing/notebooks/5.linear-least-square/

Reply:
I have changed the matrix multiply to a linear operation so that the work can be done in n^3 steps now.

Reply:
Good!